### PR TITLE
GET PATCH userGroups Information ("AccountTypes")

### PR DESCRIPTION
**Pull request by patelabhishek9893**
This commit enhances the redfish API to retrieve and set userGroups
information for each user account.

"Redfish" is always enabled in each user role. That's why it was
hardcoded into JSON response data in the old redfish API, where now it
gets retrieved using dbus interface xyz.openbmc_project.User.Attributes.
UserGroups.

UserGroups retrieve data are "redfish", "ssh", "web", and "ipmi", where
redfish DMTF Schema has predefined enum type as described below.

  "AccountTypes": {
    "enum":["Redfish", "SNMP", "OEM", "HostConsole", "ManagerConsole",
            "IPMI", "KVMIP", "VirtualMedia", "WebUI"]
   }

Here UserGroups ssh is mapped with two AccountTypes "HostConsole",
"ManagerConsole".

  - Redfish ManagerConsole == SSH to port 22 == Phosphor User manager
    Phosphor “ssh” privilege
  - Redfish HostConsole == SSH to port 2200 (host console), Which
    OpenBMC implements using the Phosphor User manager Phosphor “ssh”
    privilege.

This commit also enhances the redfish API to set and unset userGroups
information for each user account.

Users with ConfigureUsers level privilege can patch (Set and Unset)
AccountTypes of each user role. In addition, a user with
"ConfigureSelf" level privilege can only set or Update their password.

"Redfish" is always enabled in each user role. However,
"ConfigureUsers" can disable other user redfish services. But if
"ConfigureUsers" try to disable its redfish service, that generates an
error.

In this commit, users can enable and disable "redfish", "ssh", and
"ipmi" services from each user.

Tested: tested using the curl command. test using both privileges
"ConfigureUsers", "ConfigureSelf"

curl -k -H "X-Auth-Token: $token" -X PATCH [https://${bmc}/redfish/v1/](https://%24%7Bbmc%7D/redfish/v1/)
  AccountService/Accounts/operator_user -d '{"AccountTypes":["IPMI",
  "HostConsole","ManagerConsole","ManagerConsole"]}'
  
curl -k -H "X-Auth-Token: $token" -X GET [https://${bmc}/redfish/v1/](https://%24%7Bbmc%7D/redfish/v1/)
AccountService/Accounts/admin
{
  "@[odata.id](http://odata.id/)": "/redfish/v1/AccountService/Accounts/admin",
  "@odata.type": "#ManagerAccount.v1_4_0.ManagerAccount",
  "AccountTypes": [
    "IPMI",
    "Redfish",
    "HostConsole",
    "ManagerConsole"
  ],
  "Description": "User Account",
  "Enabled": true,
  "Id": "admin",
  "Links": {
    "Role": {
      "@[odata.id](http://odata.id/)": "/redfish/v1/AccountService/Roles/Administrator"
    }
  },

    ...
    ...
}

upstream:
https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/50835
https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/50965

### CODE REVIEW

**Comment by gtmills** [comment]
@patelabhishek9893 Can you add the upstream commit?

**Comment by patelabhishek9893** [comment]
upstream:
https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/50835
https://gerrit.openbmc-project.xyz/c/openbmc/bmcweb/+/50965

**Comment by patelabhishek9893** [comment]
> @patelabhishek9893 Can you add the upstream commit?

done

File: redfish-core/lib/account_service.hpp [hunk 1]
```diff
@@ -130,6 +132,190 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+        userGroup.emplace_back("ipmi");
+    }
+    if (isWebUI)
+    {
+        userGroup.emplace_back("web");
+    }
+    if ((isHostConsole) ^ (isManagerConsole))
+    {
+        BMCWEB_LOG_ERROR << "HostConsole or ManagerConsole, one of value is "
+                            "missing to set SSH property";
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:225]: You could send Redfish message StrictAccountTypes for this condition.
https://github.com/openbmc/bmcweb/blob/master/redfish-core/include/registries/base_message_registry.hpp
Background: https://github.com/openbmc/bmcweb/blob/master/static/redfish/v1/schema/ManagerAccount_v1.xml

File: redfish-core/lib/account_service.hpp [hunk 2]
```diff
@@ -130,6 +132,190 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+    // user can not disable their own Redfish Property.
+    if (isUserItself)
+    {
+        if (auto it = std::find(accountType->cbegin(), accountType->cend(),
+                                "Redfish");
+            it == accountType->cend())
+        {
+            BMCWEB_LOG_ERROR
+                << "user can not disable their own Redfish Property";
+            messages::accountNotModified(asyncResp->res);
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:274]: This should send the StrictAccountTypes message.  (See comment above for details.)
We'll have to document that: IBM systems have restrictions on AccountTypes:
1. A user account AccountTypes is required to contain the "Redfish" value.
2. For accounts which have the Administrator role, the AccountTypes is required to contain the "ManagerConsole" and "HostConsole" values.

File: redfish-core/lib/account_service.hpp [hunk 3]
```diff
@@ -130,6 +132,190 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+            return;
+        }
+    }
+
+    // MAP userGroup with accountTypes value
+    std::vector<std::string> updatedUserGroup;
+    if (!getUserGroupFromAccountType(accountType, updatedUserGroup))
+    {
+        BMCWEB_LOG_ERROR << "accountType value unable to mapped";
+        messages::internalError(asyncResp->res);
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:284]: This message is not exactly correct.  (It is *mostly* correct.)  The condition
((isHostConsole) ^ (isManagerConsole))
should result in the StrictAccountTypes message, not an internalError.

Also, if someone PATCHes an unexpected value like {"AccountTypes": ["Redfish", "BadValue"]} they should get the PropertyValueNotInList message, not an internalError.

File: redfish-core/lib/account_service.hpp [hunk 4]
```diff
@@ -130,6 +132,190 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+    // ssh configuration can't disable for Admin User
+    if (isAdminUser)
+    {
+        if (auto it = std::find(updatedUserGroup.cbegin(),
+                                updatedUserGroup.cend(), "ssh");
+            it == updatedUserGroup.cend())
+        {
+            BMCWEB_LOG_ERROR
+                << "ssh configuration can't disable for Admin User";
+            messages::accountNotModified(asyncResp->res);
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:297]: This should send the StrictAccountTypes message

**Review by joseph-reynolds — COMMENTED** [review]
The function looks good.  Thank you!  It should send better response messages.

File: redfish-core/lib/account_service.hpp (hunk 1 above)
Comment (gtmills) [reply redfish-core/lib/account_service.hpp:225]: https://redfish.dmtf.org/registries/Base.1.12.1.json
```
       "StrictAccountTypes": {
            "Description": "Indicates the request failed because a set of `AccountTypes` or `OEMAccountTypes` was not accepted while `StrictAccountTypes` is set to `true`.",
            "LongDescription": "This message shall be used to indicate that the request failed because a set of `AccountTypes` or `OEMAccountTypes` was not accepted while `StrictAccountTypes` is set to `true`.",
            "Message": "The request was not possible to fulfill with the account types included in property '%1' and property StrictAccountTypes set to true.",
            "Severity": "Warning",
            "MessageSeverity": "Warning",
            "VersionAdded": "1.9.0",
            "NumberOfArgs": 1,
            "ParamTypes": [
                "string"
            ],
            "ArgDescriptions": [
                "The name of the property."
            ],
            "ArgLongDescriptions": [
                "This argument shall contain the name of the property `AccountTypes` or `OEMAccountTypes`."
            ],
            "Resolution": "Resubmit the request either with an acceptable set of AccountTypes and OEMAccountTypes or with StrictAccountTypes set to false."
        },
```

We should also put out that StrictAccountTypes and just hardcode to True since we are enforcing it

File: redfish-core/lib/account_service.hpp (hunk 4 above)
Comment (gtmills) [reply redfish-core/lib/account_service.hpp:297]: I don't think StrictAccountTypes  makes sense here.. How about PropertyNotWritable ?

File: redfish-core/lib/account_service.hpp [hunk 5]
```diff
@@ -1374,6 +1560,40 @@ inline void updateUserProperties(std::shared_ptr<bmcweb::AsyncResp> asyncResp,
...
+                        {
+                            BMCWEB_LOG_ERROR << "UserPrivilege wasn't a "
+                                                "string";
+                            messages::internalError(asyncResp->res);
+                            return;
+                        }
+
+                        // check is user is admin user or not
+                        bool isAdminUser =
+                            (*userPrivPtr == "priv-admin") ? true : false;
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:1587]: We typically don't use the ternary operator in bmcweb

File: redfish-core/lib/account_service.hpp [hunk 6]
```diff
@@ -105,6 +106,7 @@ inline std::string getRoleIdFromPrivilege(std::string_view role)
     }
     return "";
 }
+
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:109]: be careful adding empty lines, makes reviewing harder

File: redfish-core/lib/account_service.hpp (hunk 2 above)
Comment (patelabhishek9893) [reply redfish-core/lib/account_service.hpp:274]: Agree @joseph-reynolds  what do you say?

File: redfish-core/lib/account_service.hpp (hunk 6 above)
Comment (patelabhishek9893) [reply redfish-core/lib/account_service.hpp:109]: ok, I like to add space between two function

File: redfish-core/lib/account_service.hpp (hunk 1 above)
Comment (joseph-reynolds) [reply redfish-core/lib/account_service.hpp:225]: Agreed.  Change GET /redfish/v1/AccountService property {"StrictAccountTypes": true}

File: redfish-core/lib/account_service.hpp (hunk 2 above)
Comment (joseph-reynolds) [reply redfish-core/lib/account_service.hpp:274]: I disagree.  The AccountTypes property is writable, so the PropertyNotWritable message is not appropriate.

The StrictAccountTypes message is intended to inform the caller there was a problem with the AccountTypes they specified.  This will drive them to the docs to find out what is wrong and how to fix it.  Our docs will then tell them admins must have the "ManagerConsole" and "HostConsole" values.  <--- That is what we discussed at the Redfish forum meeting when we added StrictAccountTypes to the spec.

If the AccountTypes is incorrect, the request must fail.

I will ask our information development (docs) team to add this to our BMC account management section.

File: redfish-core/lib/account_service.hpp (hunk 4 above)
Comment (joseph-reynolds) [reply redfish-core/lib/account_service.hpp:297]: I disagree.  See above.

File: redfish-core/lib/account_service.hpp (hunk 5 above)
Comment (joseph-reynolds) [reply redfish-core/lib/account_service.hpp:1587]: I tried to use the ?: operator once....
Can we use `bool isAdminUser = (*userPrivPtr == "priv-admin")`?
But we might not need this.  See my comment above where this is used.

File: redfish-core/lib/account_service.hpp [hunk 7]
```diff
@@ -130,6 +132,190 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+    // MAP userGroup with accountTypes value
+    std::vector<std::string> updatedUserGroup;
+    if (!getUserGroupFromAccountType(accountType, updatedUserGroup))
+    {
+        BMCWEB_LOG_ERROR << "accountType value unable to mapped";
+        messages::internalError(asyncResp->res);
+        return;
+    }
+
+    // ssh configuration can't disable for Admin User
```
[the remark this answers is not in the document]
Comment (patelabhishek9893) [reply redfish-core/lib/account_service.hpp:299]: do I need to remove this code?

File: redfish-core/lib/account_service.hpp [hunk 8]
```diff
@@ -130,6 +132,187 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+    }
+
+    if ((isHostConsole) ^ (isManagerConsole))
+    {
+        BMCWEB_LOG_ERROR << "HostConsole or ManagerConsole, one of value is "
+                            "missing to set SSH property";
+        isFoundAccountTypes = false;
+        if (!isHostConsole)
+        {
+            messages::strictAccountTypes(asyncResp->res, "HostConsole");
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:220]: The message replacement field is a property name.  Please change the replacement string to "AccountTypes".
The name of the AccountType value which caused this message to be sent, is not part of the message.

File: redfish-core/lib/account_service.hpp [hunk 9]
```diff
@@ -130,6 +132,187 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+        }
+        if (!isManagerConsole)
+        {
+            messages::strictAccountTypes(asyncResp->res, "ManagerConsole");
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:224]: Please change the replacement string to "AccountTypes".

File: redfish-core/lib/account_service.hpp [hunk 10]
```diff
@@ -130,6 +132,187 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+    // user can not disable their own Redfish Property.
+    if (isUserItself)
+    {
+        if (auto it = std::find(accountType->cbegin(), accountType->cend(),
+                                "Redfish");
+            it == accountType->cend())
+        {
+            BMCWEB_LOG_ERROR
+                << "user can not disable their own Redfish Property";
+            messages::strictAccountTypes(asyncResp->res, "Redfish");
```
Comment (joseph-reynolds) [inline redfish-core/lib/account_service.hpp:286]: Please change the replacement string to "AccountTypes".

**Review by joseph-reynolds — COMMENTED** [review]
Looks good to me, except for the parameter for the StrictAccountTypes message.

**Review by joseph-reynolds — APPROVED** [review]
Looks good, thank you!

File: redfish-core/include/error_messages.hpp [hunk 11]
```diff
@@ -608,6 +608,17 @@ nlohmann::json sourceDoesNotSupportProtocol(const std::string& arg1,
 void sourceDoesNotSupportProtocol(crow::Response& res, const std::string& arg1,
                                   const std::string& arg2);
 
+/**
+ * @brief Formats StrictAccountTypes message into JSON
+ * Message body: Indicates the request failed because a set of `AccountTypes` or
+ * `OEMAccountTypes` was not accepted while `StrictAccountTypes` is set to `true
+ * @param[in] arg1 Parameter of message that will replace %1 in its body.
```
Comment (aahmed-2) [inline redfish-core/include/error_messages.hpp:615]: Maybe rephrase? 

What does %1 mean? I don't see %1 in error_message.cpp definitions of strictAccountTypes().

File: redfish-core/lib/account_service.hpp [hunk 12]
```diff
@@ -130,6 +132,187 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+    else if (userGroup == "ipmi")
+    {
+        accountTypes.push_back("IPMI");
+    }
+    else if (userGroup == "ssh")
+    {
+        accountTypes.push_back("HostConsole");
+        accountTypes.push_back("ManagerConsole");
+    }
+    else if (userGroup == "web")
```
Comment (aahmed-2) [inline redfish-core/lib/account_service.hpp:154]: Will userGroup be set to "web" for phosphor-webui and webui-vue?

File: redfish-core/lib/account_service.hpp [hunk 13]
```diff
@@ -130,6 +132,187 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+        {
+            // set false if accountTypes not found and return
+            isFoundAccountTypes = false;
+            messages::propertyValueNotInList(asyncResp->res, "AccountTypes",
+                                             accountType);
+            return isFoundAccountTypes;
+        }
+    }
+
+    if ((isHostConsole) ^ (isManagerConsole))
```
Comment (aahmed-2) [inline redfish-core/lib/account_service.hpp:213]: Just double checking the desired logic here is that 
BOTH hostConsole and ManagerConsole need to be true for a valid account UserGroup?

File: redfish-core/lib/account_service.hpp [hunk 14]
```diff
@@ -2129,11 +2326,14 @@ inline void requestAccountServiceRoutes(App& app)
             std::optional<std::string> roleId;
             std::optional<bool> locked;
             std::optional<nlohmann::json> oem;
+            std::optional<std::vector<std::string>> accountType;
+
+            bool isUserItself = (username == req.session->username);
```
Comment (aahmed-2) [inline redfish-core/lib/account_service.hpp:2331]: Should isUserItself be set after we check if req.session is a nullptr (as in line 2342)?

**Review by aahmed-2 — APPROVED** [review]

File: redfish-core/lib/account_service.hpp (hunk 13 above)
Comment (patelabhishek9893) [reply redfish-core/lib/account_service.hpp:213]: either both need to false or either both need to be true
so this is perfect

File: redfish-core/lib/account_service.hpp (hunk 12 above)
Comment (patelabhishek9893) [reply redfish-core/lib/account_service.hpp:154]: for both, I think
because I think at a time only one get used

File: redfish-core/include/error_messages.hpp (hunk 11 above)
Comment (patelabhishek9893) [reply redfish-core/include/error_messages.hpp:615]: %1 got replaced with arg1 in that code

**Review by aahmed-2 — APPROVED** [review]
Looks good to me!

**Review by joseph-reynolds — APPROVED** [review]
Good catch!  Approved.

File: redfish-core/lib/account_service.hpp [hunk 15]
```diff
@@ -130,6 +132,187 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+    if ((isHostConsole) ^ (isManagerConsole))
+    {
+        BMCWEB_LOG_ERROR << "HostConsole or ManagerConsole, one of value is "
+                            "missing to set SSH property";
+        isFoundAccountTypes = false;
+        if (!isHostConsole)
+        {
+            messages::strictAccountTypes(asyncResp->res, "AccountTypes");
+        }
+        if (!isManagerConsole)
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:222]: These are the same message.. Did you want to add HostConsole or ManagerConsole to this error message?

File: redfish-core/lib/account_service.hpp [hunk 16]
```diff
@@ -130,6 +132,187 @@ inline std::string getPrivilegeFromRoleId(std::string_view role)
...
+    // user can not disable their own Redfish Property.
+    if (isUserItself)
+    {
+        if (auto it = std::find(accountType->cbegin(), accountType->cend(),
+                                "Redfish");
+            it == accountType->cend())
+        {
+            BMCWEB_LOG_ERROR
+                << "user can not disable their own Redfish Property";
+            messages::strictAccountTypes(asyncResp->res, "AccountTypes");
```
Comment (gtmills) [inline redfish-core/lib/account_service.hpp:286]: This doesn't sound like the correct error message..

File: redfish-core/lib/account_service.hpp (hunk 2 above)
Comment (gtmills) [reply redfish-core/lib/account_service.hpp]: The StrictAccountTypes message is intended when StrictAccountTypes = True and a value that would be enabled because it is tied to another value being set is left out of the request.
 (e.g. HostConsole is passed and ManagerConsole isn't )
 This code is for disabling Redfish, this doesn't have anything to do with StrictAccountTypes = True (e.g. if we added write support to StrictAccountTypes and set it to false, this check and error message would still be there) 
 The StrictAccountTypes  message doesn't make sense here

**Review by gtmills — APPROVED** [review]

**Comment by joseph-reynolds** [comment]
Bumped by https://github.ibm.com/openbmc/openbmc/pull/1991
